### PR TITLE
Detect recursive method calls in lambda expressions

### PR DESCRIPTION
**Pull request by yaroslavafenkin**
This used to be based on https://github.com/jenkins-infra/usage-in-plugins/pull/29.
Relevant diff used to be in https://github.com/yaroslavafenkin/usage-in-plugins/compare/recursive...lambdas.

After https://github.com/jenkins-infra/usage-in-plugins/pull/29 got merged it's now easier to see what this PR adds :)

### CODE REVIEW

**Review by dwnusbaum — COMMENTED** [review]
Looks ok (I did not review the changes from #21 in detail), but IDK if the behavior is what you want.

File: src/main/java/org/jenkinsci/deprecatedusage/DeprecatedUsage.java [hunk 1]
```diff
@@ -264,47 +289,70 @@ public void visit(int version, int access, String name, String signature, String
...
-            methodCalled(owner, name, desc);
+            methodCalled(owner, name, desc, this.className, this.name, this.desc);
+        }
+
+        @Override
+        public void visitInvokeDynamicInsn(String name, String descriptor, Handle bootstrapMethodHandle, 
+                                           Object... bootstrapMethodArguments) {
+            Handle methodArgument = (Handle) bootstrapMethodArguments[1];
+            methodCalled(methodArgument.getOwner(), methodArgument.getName(), methodArgument.getDesc(), 
+                    this.className, this.name, this.desc);
```
[the remark this answers is not in the document]
Comment (daniel-beck) [reply src/main/java/org/jenkinsci/deprecatedusage/DeprecatedUsage.java:350]: Related?

```
Failed to fully analyze .../usage-in-plugins/work/instance-identity/116.vf8f487400980/instance-identity.hpi.  org/jenkinsci/main/modules/instance_identity/InstanceIdentity.class not scanned due to Index 1 out of bounds for length 1
Failed to fully analyze .../usage-in-plugins/work/instance-identity/116.vf8f487400980/instance-identity.hpi.  org/jenkinsci/main/modules/instance_identity/pem/DerParser.class not scanned due to Index 1 out of bounds for length 1
Failed to fully analyze .../usage-in-plugins/work/instance-identity/116.vf8f487400980/instance-identity.hpi.  org/jenkinsci/main/modules/instance_identity/SelfSignedCertificate.class not scanned due to Index 1 out of bounds for length 1
Failed to fully analyze .../usage-in-plugins/work/custom-folder-icon/2.0/custom-folder-icon.hpi.  jenkins/plugins/foldericon/BuildStatusFolderIcon.class not scanned due to Index 1 out of bounds for length 1
Failed to fully analyze .../usage-in-plugins/work/custom-folder-icon/2.0/custom-folder-icon.hpi.  jenkins/plugins/foldericon/CustomFolderIcon$DescriptorImpl.class not scanned due to Index 1 out of bounds for length 1
Failed to fully analyze .../usage-in-plugins/work/custom-folder-icon/2.0/custom-folder-icon.hpi.  jenkins/plugins/foldericon/CustomFolderIcon.class not scanned due to Index 1 out of bounds for length 1

[the remark this answers is not in the document]
Comment (daniel-beck) [reply src/main/java/org/jenkinsci/deprecatedusage/DeprecatedUsage.java:350]: Confirmed, if I print the top stack trace element, it's

```
 org.jenkinsci.deprecatedusage.DeprecatedUsage$CallersMethodVisitor.visitInvokeDynamicInsn(DeprecatedUsage.java:348)

[the remark this answers is not in the document]
Comment (daniel-beck) [reply src/main/java/org/jenkinsci/deprecatedusage/DeprecatedUsage.java:350]: > Also, for what it's worth, I think that the only new effect of this method will be to detect method calls that happen as a result of method references

FWIW I compared outputs, and if a `#lambda$whatever$X` appears anywhere in the recursive tree, it's always the top-level elements without this change, but can be in the middle ("called") with this change.

[the remark this answers is not in the document]
Comment (daniel-beck) [reply src/main/java/org/jenkinsci/deprecatedusage/DeprecatedUsage.java:350]: Noting that there's also tons of NPEs when passing `--includePluginLibs`.

[the remark this answers is not in the document]
Comment (yaroslavafenkin) [reply src/main/java/org/jenkinsci/deprecatedusage/DeprecatedUsage.java:350]: > Interesting, I wonder what is special about those cases.

Only pointer I have so far is [StringConcatFactory.html#makeConcatWithConstants](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/invoke/StringConcatFactory.html#makeConcatWithConstants(java.lang.invoke.MethodHandles.Lookup,java.lang.String,java.lang.invoke.MethodType,java.lang.String,java.lang.Object...)).

I am not able to describe what exactly happens in those cases without spending a dozen of hours on it.

[the remark this answers is not in the document]
Comment (yaroslavafenkin) [reply src/main/java/org/jenkinsci/deprecatedusage/DeprecatedUsage.java:350]: > Noting that there's also tons of NPEs when passing `--includePluginLibs`.

NPEs of what kind? I only see things like:
```
....Failed to fully analyze /usage-in-plugins/work/deploy/1.16/deploy.hpi.  module-info.class not scanned due to null
Failed to fully analyze /usage-in-plugins/work/deploy/1.16/deploy.hpi.  module-info.class not scanned due to Index 4 out of bounds for length 4
```
```
Failed to fully analyze /usage-in-plugins/work/graphiteIntegrator/1.2/graphiteIntegrator.hpi.  com/ibm/icu/impl/data/LocaleElements_zh__PINYIN.class not scanned due to null
```
```
Failed to fully analyze /usage-in-plugins/work/discard-old-build/1.05/discard-old-build.hpi.  com/ibm/icu/impl/data/LocaleElements_zh__PINYIN.class not scanned due to null
```

**Review by daniel-beck — APPROVED** [review]
Works very well in local testing 👍

File: src/main/java/org/jenkinsci/deprecatedusage/DeprecatedUsage.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (daniel-beck) [reply src/main/java/org/jenkinsci/deprecatedusage/DeprecatedUsage.java]: > After applying Devin's suggestion I only see things like

Those are the ones I am currently still getting too. There's also:

```
Failed to fully analyze /Users/danielbeck/Repositories/github.com/jenkins-infra/usage-in-plugins/work/deploy/1.16/deploy.hpi.  module-info.class not scanned due to 32800
Failed to fully analyze /Users/danielbeck/Repositories/github.com/jenkins-infra/usage-in-plugins/work/deploy/1.16/deploy.hpi.  module-info.class not scanned due to 4
Failed to fully analyze /Users/danielbeck/Repositories/github.com/jenkins-infra/usage-in-plugins/work/deploy/1.16/deploy.hpi.  module-info.class not scanned due to null
Failed to fully analyze /Users/danielbeck/Repositories/github.com/jenkins-infra/usage-in-plugins/work/deploy/1.16/deploy.hpi.  module-info.class not scanned due to 4
```

which can probably also be ignored.

**Pull request merged**